REGRESSION(284072@main): [GStreamer] A lot of media tests crash with Assertion "SHOULD NEVER BE REACHED"

eocanha opened the pull request:
#### ae4d384da8afe6b90b4155f74110fd6da02dae9a
<pre>
REGRESSION(284072@main): [GStreamer] A lot of media tests crash with Assertion &quot;SHOULD NEVER BE REACHED&quot;
<a href="https://bugs.webkit.org/show_bug.cgi?id=280334">https://bugs.webkit.org/show_bug.cgi?id=280334</a>

Reviewed by Xabier Rodriguez-Calvar and Philippe Normand.

A video can work in two modes: streaming mode (GstQueue2 element is
used, GST_MESSAGE_BUFFERING is relied upon to get buffering info) or
download mode (GstDownloadBuffer element is used, fillTimer is used to
poll its buffering percentage).

However, under some circumstances (file or mediastream URI, for
instance), none of these two GStreamer elements are created and the
player private was expecting the video to work in download mode. The
failing assert higlighted the mismatch when fillTimerFired asked for the
buffering percentage. But the truth is that in these special cases the
video can actually work in stream mode and ask the whole pipeline or the
sinks for the buffering percentage (which usually is just 100%, meaning
that buffering is complete).

This commit acknowledges the existence of these special cases and
removes the assert. Also, it stops the fillTimer when buffering reaches
100% in those cases, to avoid wasting resources.

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::queryBufferingPercentage): Removed assert. Added extra comment. Stop fillTimer (and report it in the logs) when on stream mode and buffering reaching 100%.

Canonical link: <a href="https://commits.webkit.org/284353@main">https://commits.webkit.org/284353@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d6f52d8cc0e7de1aac23b4d0c3d6afee653f2bb3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/69018 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48418 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21690 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/73098 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/20176 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/71135 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/56219 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/20024 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/54958 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13403 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/72084 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44201 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59584 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35432 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40866 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/17013 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18550 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62815 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17359 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74807 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/13000 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16609 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/62602 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/13038 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59667 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62499 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15340 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10484 "Passed tests") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4091 "Found 1 new test failure: http/tests/media/media-source/media-source-video-fit-fill.html (failure)") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/44222 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/45295 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46491 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/45037 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

calvaris reviewed (approved):
I assume you tested already mediastream and things are ok.

eocanha committed:
REGRESSION(284072@main): [GStreamer] A lot of media tests crash with Assertion "SHOULD NEVER BE REACHED"
https://bugs.webkit.org/show_bug.cgi?id=280334

Reviewed by Xabier Rodriguez-Calvar and Philippe Normand.

A video can work in two modes: streaming mode (GstQueue2 element is
used, GST_MESSAGE_BUFFERING is relied upon to get buffering info) or
download mode (GstDownloadBuffer element is used, fillTimer is used to
poll its buffering percentage).

However, under some circumstances (file or mediastream URI, for
instance), none of these two GStreamer elements are created and the
player private was expecting the video to work in download mode. The
failing assert higlighted the mismatch when fillTimerFired asked for the
buffering percentage. But the truth is that in these special cases the
video can actually work in stream mode and ask the whole pipeline or the
sinks for the buffering percentage (which usually is just 100%, meaning
that buffering is complete).

This commit acknowledges the existence of these special cases and
removes the assert. Also, it stops the fillTimer when buffering reaches
100% in those cases, to avoid wasting resources.

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::queryBufferingPercentage): Removed assert. Added extra comment. Stop fillTimer (and report it in the logs) when on stream mode and buffering reaching 100%.

Canonical link: https://commits.webkit.org/284353@main
